fdbserver/storageserver.actor.cpp

```diff
@@ -1204,6 +1211,37 @@ Future<Version> waitForVersion(StorageServer* data, Version version, SpanID span
 	return waitForVersionActor(data, version, spanContext);
 }
 
+Future<Version> waitForVersion(StorageServer* data, Version commitVersion, Version readVersion, SpanID spanContext) {
+	ASSERT(commitVersion == invalidVersion || commitVersion < readVersion);
```

Comment:
I wonder if `commitVersion <= readVersion`?

Reply:
I don't think so. This is because DatabaseContext::getLatestCommitVersions() adds a tag to the version vector only if "commitVersion < readVersion":

```
void DatabaseContext::getLatestCommitVersions(const Reference<LocationInfo>& locationInfo,
                                              Version readVersion,
                                              VersionVector& latestCommitVersions) {
	std::map<Version, std::set<Tag>> versionMap; // order the versions to be returned
	for (int i = 0; i < locationInfo->locations()->size(); i++) {
		UID uid = locationInfo->locations()->getId(i);
		if (ssidTagMapping.find(uid) != ssidTagMapping.end()) {
			Tag tag = ssidTagMapping[uid];
			if (ssVersionVectorCache.hasVersion(tag)) {
				Version commitVersion = ssVersionVectorCache.getVersion(tag); // latest commit version
				if (commitVersion < readVersion) {
					versionMap[commitVersion].insert(tag);
				}
			}
		}
	}
```

Reply:
Interesting. Why `getLatestCommitVersions()` behaves this way? Isn't it possible that `commitVersion == readVersion`, i.e., the tag/SS has latest commits == the read version?

Reply:
Yes, it is possible. There is no need to propagate the commit version in that case, right? That is, the storage server will be doing a read at the read version (= commit version) anyways. We can propagate the commit version too, there is no harm in doing that, but it wouldn't provide any additional information to the storage server and so didn't do it.